// @mui/x-data-grid/legacy/hooks/features/editRows/useGridCellEditing.new.js
import _asyncToGenerator from "@babel/runtime/helpers/esm/asyncToGenerator";
import _defineProperty from "@babel/runtime/helpers/esm/defineProperty";
import _extends from "@babel/runtime/helpers/esm/extends";
import _regeneratorRuntime from "@babel/runtime/regenerator";
import * as React from 'react';
import { useGridApiEventHandler, useGridApiOptionHandler } from '../../utils/useGridApiEventHandler';
import { GridEvents } from '../../../models/events/gridEvents';
import { GridEditModes, GridCellModes } from '../../../models/gridEditRowModel';
import { useGridApiMethod } from '../../utils/useGridApiMethod';
import { gridEditRowsStateSelector } from './gridEditRowsSelector';
import { isPrintableKey } from '../../../utils/keyboardUtils';
import { GridCellEditStartReasons, GridCellEditStopReasons } from '../../../models/params/gridEditCellParams';
export var useGridCellEditing = function useGridCellEditing(apiRef, props) {
  var processRowUpdate = props.processRowUpdate;

  var runIfEditModeIsCell = function runIfEditModeIsCell(callback) {
    return function () {
      if (props.editMode === GridEditModes.Cell) {
        callback.apply(void 0, arguments);
      }
    };
  };

  var throwIfNotEditable = React.useCallback(function (id, field) {
    var params = apiRef.current.getCellParams(id, field);

    if (!apiRef.current.isCellEditable(params)) {
      throw new Error("MUI: The cell with id=".concat(id, " and field=").concat(field, " is not editable."));
    }
  }, [apiRef]);
  var throwIfNotInMode = React.useCallback(function (id, field, mode) {
    if (apiRef.current.getCellMode(id, field) !== mode) {
      throw new Error("MUI: The cell with id=".concat(id, " and field=").concat(field, " is not in ").concat(mode, " mode."));
    }
  }, [apiRef]);
  var handleCellDoubleClick = React.useCallback(function (params, event) {
    if (!params.isEditable) {
      return;
    }

    if (params.cellMode === GridCellModes.Edit) {
      return;
    }

    var newParams = _extends({}, params, {
      reason: GridCellEditStartReasons.cellDoubleClick
    });

    apiRef.current.publishEvent(GridEvents.cellEditStart, newParams, event);
  }, [apiRef]);
  var handleCellFocusOut = React.useCallback(function (params, event) {
    if (params.cellMode === GridCellModes.View) {
      return;
    }

    var newParams = _extends({}, params, {
      reason: GridCellEditStopReasons.cellFocusOut
    });

    apiRef.current.publishEvent(GridEvents.cellEditStop, newParams, event);
  }, [apiRef]);
  var handleCellKeyDown = React.useCallback(function (params, event) {
    if (params.cellMode === GridCellModes.Edit) {
      var reason;

      if (event.key === 'Escape') {
        reason = GridCellEditStopReasons.escapeKeyDown;
      } else if (event.key === 'Enter') {
        reason = GridCellEditStopReasons.enterKeyDown;
      } else if (event.key === 'Tab') {
        reason = event.shiftKey ? GridCellEditStopReasons.shiftTabKeyDown : GridCellEditStopReasons.tabKeyDown;
        event.preventDefault(); // Prevent going to the next element in the tab sequence
      }

      if (reason) {
        var newParams = _extends({}, params, {
          reason: reason
        });

        apiRef.current.publishEvent(GridEvents.cellEditStop, newParams, event);
      }
    } else if (params.isEditable) {
      var _reason;

      if (isPrintableKey(event.key)) {
        if (event.shiftKey || event.ctrlKey || event.metaKey || event.altKey) {
          return;
        }

        _reason = GridCellEditStartReasons.printableKeyDown;
      } else if (event.key === 'Enter') {
        _reason = GridCellEditStartReasons.enterKeyDown;
      } else if (event.key === 'Delete') {
        _reason = GridCellEditStartReasons.deleteKeyDown;
      }

      if (_reason) {
        var _newParams = _extends({}, params, {
          reason: _reason
        });

        apiRef.current.publishEvent(GridEvents.cellEditStart, _newParams, event);
      }
    }
  }, [apiRef]);
  var handleCellEditStart = React.useCallback(function (params, event) {
    var id = params.id,
        field = params.field,
        reason = params.reason;
    apiRef.current.startCellEditMode(params);

    if (reason === GridCellEditStartReasons.deleteKeyDown || reason === GridCellEditStartReasons.printableKeyDown) {
      apiRef.current.setEditCellValue({
        id: id,
        field: field,
        value: ''
      }, event);
    }
  }, [apiRef]);
  var handleCellEditStop = React.useCallback(function (params) {
    var id = params.id,
        field = params.field,
        reason = params.reason;
    var cellToFocusAfter;

    if (reason === GridCellEditStopReasons.enterKeyDown) {
      cellToFocusAfter = 'below';
    } else if (reason === GridCellEditStopReasons.tabKeyDown) {
      cellToFocusAfter = 'right';
    } else if (reason === GridCellEditStopReasons.shiftTabKeyDown) {
      cellToFocusAfter = 'left';
    }

    var ignoreModifications = reason === 'escapeKeyDown';
    var editingState = gridEditRowsStateSelector(apiRef.current.state);

    if (editingState[id][field].isProcessingProps) {
      // The user wants to stop editing the cell but we can't wait for the props to be processed.
      // In this case, discard the modifications.
      ignoreModifications = true;
    }

    apiRef.current.stopCellEditMode({
      id: id,
      field: field,
      ignoreModifications: ignoreModifications,
      cellToFocusAfter: cellToFocusAfter
    });
  }, [apiRef]);
  useGridApiEventHandler(apiRef, GridEvents.cellDoubleClick, runIfEditModeIsCell(handleCellDoubleClick));
  useGridApiEventHandler(apiRef, GridEvents.cellFocusOut, runIfEditModeIsCell(handleCellFocusOut));
  useGridApiEventHandler(apiRef, GridEvents.cellKeyDown, runIfEditModeIsCell(handleCellKeyDown));
  useGridApiEventHandler(apiRef, GridEvents.cellEditStart, runIfEditModeIsCell(handleCellEditStart));
  useGridApiEventHandler(apiRef, GridEvents.cellEditStop, runIfEditModeIsCell(handleCellEditStop));
  useGridApiOptionHandler(apiRef, GridEvents.cellEditStart, props.onCellEditStart);
  useGridApiOptionHandler(apiRef, GridEvents.cellEditStop, props.onCellEditStop);
  var getCellMode = React.useCallback(function (id, field) {
    var editingState = gridEditRowsStateSelector(apiRef.current.state);
    var isEditing = editingState[id] && editingState[id][field];
    return isEditing ? GridCellModes.Edit : GridCellModes.View;
  }, [apiRef]);
  var updateOrDeleteFieldState = React.useCallback(function (id, field, newProps) {
    apiRef.current.setState(function (state) {
      var newEditingState = _extends({}, state.editRows);

      if (newProps !== null) {
        newEditingState[id] = _extends({}, newEditingState[id], _defineProperty({}, field, _extends({}, newProps)));
      } else {
        delete newEditingState[id][field];

        if (Object.keys(newEditingState[id]).length === 0) {
          delete newEditingState[id];
        }
      }

      return _extends({}, state, {
        editRows: newEditingState
      });
    });
    apiRef.current.forceUpdate();
  }, [apiRef]);
  var startCellEditMode = React.useCallback(function (params) {
    var id = params.id,
        field = params.field;
    throwIfNotEditable(id, field);
    throwIfNotInMode(id, field, GridCellModes.View);
    var newProps = {
      value: apiRef.current.getCellValue(id, field),
      error: false,
      isProcessingProps: false
    };
    updateOrDeleteFieldState(id, field, newProps);
    apiRef.current.setCellFocus(id, field);
  }, [apiRef, throwIfNotEditable, throwIfNotInMode, updateOrDeleteFieldState]);
  var stopCellEditMode = React.useCallback(function (params) {
    var id = params.id,
        field = params.field,
        ignoreModifications = params.ignoreModifications,
        _params$cellToFocusAf = params.cellToFocusAfter,
        cellToFocusAfter = _params$cellToFocusAf === void 0 ? 'none' : _params$cellToFocusAf;
    throwIfNotInMode(id, field, GridCellModes.Edit);
    apiRef.current.unstable_runPendingEditCellValueMutation(id, field);

    var updateFocusedCellIfNeeded = function updateFocusedCellIfNeeded() {
      if (cellToFocusAfter !== 'none') {
        apiRef.current.unstable_moveFocusToRelativeCell(id, field, cellToFocusAfter);
      }
    };

    if (ignoreModifications) {
      updateFocusedCellIfNeeded();
      updateOrDeleteFieldState(id, field, null);
      return;
    }

    var editingState = gridEditRowsStateSelector(apiRef.current.state);
    var row = apiRef.current.getRow(id);
    var column = apiRef.current.getColumn(field);
    var _editingState$id$fiel = editingState[id][field],
        value = _editingState$id$fiel.value,
        error = _editingState$id$fiel.error,
        isProcessingProps = _editingState$id$fiel.isProcessingProps;

    if (error || isProcessingProps) {
      return;
    }

    var rowUpdate = column.valueSetter ? column.valueSetter({
      value: value,
      row: row
    }) : _extends({}, row, _defineProperty({}, field, value));

    if (processRowUpdate) {
      Promise.resolve(processRowUpdate(rowUpdate, row)).then(function (finalRowUpdate) {
        apiRef.current.updateRows([finalRowUpdate]);
        updateFocusedCellIfNeeded();
        updateOrDeleteFieldState(id, field, null);
      });
    } else {
      apiRef.current.updateRows([rowUpdate]);
      updateFocusedCellIfNeeded();
      updateOrDeleteFieldState(id, field, null);
    }
  }, [apiRef, processRowUpdate, throwIfNotInMode, updateOrDeleteFieldState]);
  var setCellEditingEditCellValue = React.useCallback( /*#__PURE__*/function () {
    var _ref = _asyncToGenerator( /*#__PURE__*/_regeneratorRuntime.mark(function _callee(params) {
      var id, field, value, column, row, parsedValue, editingState, newProps, hasChanged;
      return _regeneratorRuntime.wrap(function _callee$(_context) {
        while (1) {
          switch (_context.prev = _context.next) {
            case 0:
              id = params.id, field = params.field, value = params.value;
              throwIfNotEditable(id, field);
              throwIfNotInMode(id, field, GridCellModes.Edit);
              column = apiRef.current.getColumn(field);
              row = apiRef.current.getRow(id);
              parsedValue = value;

              if (column.valueParser) {
                parsedValue = column.valueParser(value, apiRef.current.getCellParams(id, field));
              }

              editingState = gridEditRowsStateSelector(apiRef.current.state);
              newProps = _extends({}, editingState[id][field], {
                value: parsedValue
              });

              if (!column.preProcessEditCellProps) {
                _context.next = 16;
                break;
              }

              hasChanged = value !== editingState[id][field].value;
              newProps = _extends({}, newProps, {
                isProcessingProps: true
              });
              updateOrDeleteFieldState(id, field, newProps);
              _context.next = 15;
              return Promise.resolve(column.preProcessEditCellProps({
                id: id,
                row: row,
                props: newProps,
                hasChanged: hasChanged
              }));

            case 15:
              newProps = _context.sent;

            case 16:
              if (!(apiRef.current.getCellMode(id, field) === GridCellModes.View)) {
                _context.next = 18;
                break;
              }

              return _context.abrupt("return", false);

            case 18:
              editingState = gridEditRowsStateSelector(apiRef.current.state);
              newProps = _extends({}, newProps, {
                isProcessingProps: false
              }); // We don't update the value with the one coming from the props pre-processing
              // because when the promise resolves it may be already outdated. The only
              // exception to this rule is when there's no pre-processing.

              newProps.value = column.preProcessEditCellProps ? editingState[id][field].value : parsedValue;
              updateOrDeleteFieldState(id, field, newProps);
              editingState = gridEditRowsStateSelector(apiRef.current.state);
              return _context.abrupt("return", !editingState[id][field].error);

            case 24:
            case "end":
              return _context.stop();
          }
        }
      }, _callee);
    }));

    return function (_x) {
      return _ref.apply(this, arguments);
    };
  }(), [apiRef, throwIfNotEditable, throwIfNotInMode, updateOrDeleteFieldState]);
  var editingApi = {
    getCellMode: getCellMode,
    startCellEditMode: startCellEditMode,
    stopCellEditMode: stopCellEditMode,
    unstable_setCellEditingEditCellValue: setCellEditingEditCellValue
  };
  useGridApiMethod(apiRef, editingApi, 'EditingApi');
};
